### ISSUE TITLE
AdvancedUserInterface is not supported

### ISSUE THREAD

**Issue by micronax** [scheb/two-factor-bundle#206]
I currently face some problems while implementing the 2FA in a project:

I followed all steps from the documentation and was successfully able to enable 2FA for a testuser with Google Authenticator.

After login I get redirected to /2fa (default configuration). When I successfully pass the 2FA-challenge (I get `User has been two-factor authenticated successfully` in the logs) I get redirected to the default route behind the firewall. Then I get redirected to the login form again because of insufficient role permissions. When I re-login or directly point to a route behind the firewall I get redirected again to /2fa.

In the symfony debug toolbar I see, I have a TwoFactorToken with a UsernamePasswordToken embedded which contains the correct roles. But the TwoFactorToken itself seems like containing no roles at all.

**The profile output of the /login route:**
![Bildschirmfoto 2019-04-22 um 10 12 49](https://user-images.githubusercontent.com/1337823/56490991-5b177e00-64e7-11e9-96ec-0464721c0976.png)

**Output of `dump($tokenStorage->getToken());` at the /login route:**
![Bildschirmfoto 2019-04-22 um 10 02 46](https://user-images.githubusercontent.com/1337823/56491001-623e8c00-64e7-11e9-8a97-a727bcf86887.png)

**Some more debugging output:**
![Bildschirmfoto 2019-04-22 um 10 00 19](https://user-images.githubusercontent.com/1337823/56491030-784c4c80-64e7-11e9-8b45-5f9acab4af19.png)

**My Configuration:**

**routes.yml**
```
2fa_login:
    path: /2fa
    defaults:
        _controller: "scheb_two_factor.form_controller:form"

2fa_login_check:
    path: /2fa_check
```

**config.yml**
```
scheb_two_factor:
    google:
        enabled:     true
```

**security.yml**
```
firewalls:
        # disables authentication for assets and the profiler, adapt it according to your needs
        dev:
            pattern: ^/(_(profiler|wdt)|css|images|js)/
            security: false

        # Main protected user-area
        webgui:
            pattern:   ^/
            anonymous: ~
            provider: entity_provider
            switch_user: { role: ROLE_ADMIN_SWITCH_USER, parameter: _app_switch_user }
            form_login:
                login_path: login
                check_path: login
                default_target_path: login_switch_redirect
                use_referer: true
            logout:
                path:   /logout
                target: /login
            two_factor:
                auth_form_path: 2fa_login
                check_path:     2fa_login_check
                default_target_path: /dashboard
                always_use_default_target_path: false
```

**Comment by scheb** [comment]
The second image with "TwoFactorListener.php on line 253" tells me, that you reach the point where the two-factor authentication process is completed and the authenticated token has been unwrapped from the TwoFactorToken. So the two-factor authentication did what it's supposed to do.

What is a bit suspicous is the fact, that your UsernamePasswordToken has the authenticated=false flag, which effectively tells the firewall "this token is not valid" and therefore you're redirected back to the login page. The question is, why do you end up with such a token if you did a successful login before? To debug I'd recommend:

- If you disable 2fa on the firewall, just doing a normal login, do you still end up with an authenticated=false UsernamePasswordToken?
- Do you somehow end up in one of these lines? To my understanding that's the only ways how a UsernamePasswordToken with roles could become authenticated=false. If so, check where the call is coming from.
  - https://github.com/symfony/symfony/blob/4.2/src/Symfony/Component/Security/Core/Authentication/Token/AbstractToken.php#L102
  - https://github.com/symfony/symfony/blob/4.2/src/Symfony/Component/Security/Core/Authentication/Token/UsernamePasswordToken.php#L58

**Comment by micronax** [comment]
When I remove the `two_factor` options from the firewall, authentication works as it should do.
I forgot to mention that we use **symfony 3.4.26** in this project.

During some debugging with the lines you gave me, I found out that the function `AbstractToken::hasUserChanged()` returns false with the check `$this->user->isEnabled() !== $user->isEnabled()`.

During a `dump($user, $this->user);` I found out that `$this->user` is not populated with the full user properties. Only the username and the hashed password is set. The other properties are null.

Does the 2FA-Bundle support usage of the AdvancedUsedInterface (which I think is deprecated in SF4)? Any chance to get the user object populated with all required parameters?

Thank you in advance.

**Comment by micronax** [comment]
I think Im going to migrate the AdvancedUserInterface to a UserChecker but this issue may impact other users too..

**Comment by micronax** [comment]
I can confirm that the problem only occurs when User implements the AdvancedUserInterface. I renamed the issue appropriate.

**Comment by scheb** [comment]
I still don't understand how the AdvancedUserInterface breaks it, because AdvancedUserInterface is just an extension to the normal UserInterface. Trying to reproduce the behaviour somehow. How did your User entity with AdvancedUserInterface look like?

**Comment by micronax** [comment]
I think the problem is the following:

the 2FAToken is somehow not populated with the complete user data, only with username and password. But as mentioned above, the `AbstractToken::hasUserChanged()` function call the method needs some information to be populated, in my case the `User::isEnabled()` method checks if a field called `$status` is > 0. But the field is NULL -> false -> hasUserChanged returns false -> user is not authenticated.

**Comment by scheb** [comment]
I was able to reproduce the behavior related to `AdvancedUserInterface`, it explains why you're missing field values in the user entity. The only thing still confusing me is, that you said normal login without two-factor firewall settings work for you, because even that one is broken for me. But maybe security checks work a bit differently in my Symfony 4.2 version compared to your 3.4.

The issue that I found was, the user entity implemented a serialize/deserialize interface, and it did not serialize all the relevant fields, but only basic ones (user + password). So after the token was written to the session data, in the following request the user entity was missing field values and therefore the `hasUserChanged()` check (which only happens for `AdvancedUserInterface`) returns `false`, effectively deauthorizing the token. And with a deauthorized token, you're sent to the login form.

To verify my theory, would you please check what's happening when you serialize/deserialize a user entity?
1) Does your user entity implement `\Serializable`?
2) Get a fresh user entity from the database
3) Serialize and unserialize it: `$string = serialize($user); $unserialized = unserialize($string);`
4) Is `$user` the same as `$unserialized`? => If not: That's the problem.

**Comment by daum** [comment]
@scheb Just ran into a similar situation that was described here.  I am **not** getting the authenticated => false, mine is true.  I go through get the "User has authenticated properly" then it redirects back to the route I set as a redirect, and immediately redirects back to the 2fa screen.  I am implementing the AdvancedUserInterface (using FOSUserBundle), will try to see if I can recreate a on a completely blank project so you can see too.  Just wasn't sure if you had any other troubleshooting steps than the ones here and on the help page that you might think makes sense.

**Comment by Spomky** [comment]
[AdvancedUserInterface is deprecated](https://symfony.com/blog/new-in-symfony-4-1-deprecated-the-advanceduserinterface)

**Comment by scheb** [comment]
@daum Please have a look at the [troubleshooting guide](https://github.com/scheb/two-factor-bundle/blob/master/Resources/doc/troubleshooting.md) section "Not logged in after completing two-factor authentication".

**Comment by daum** [comment]
Alrighty I think I tracked down in my case what is doing it, I don't think it is related to this ticket but let me know.

If you have ` always_authenticate_before_granting: true` so that the roles are refreshed each page load, this causes the regular two factor provider to kick in each time, which then reverts you back to the to the two factor login.  

I made a test case here: https://github.com/daum/example-twofactor-issue If you clone, set your two factor up, login, do the two factor, it will take you to /.  Refresh the page and you'll notice you are then redirected back to the 2fa page.  

I'm not sure of the best solution here, after catching it it seems pretty obvious if you require an authenticate each time before checking for grants that it will trigger the regular 2fa steps.  I think maybe just adding this to the documentation as a known gotcha?
